### PR TITLE
Fixed error hiding core task list

### LINKED ISSUE #6049: Error hiding core task list
**Describe the bug**
If the extended task list is visible, it is not possible to hide the core task list.

**To Reproduce**

1. Add [a task list item](https://github.com/woocommerce/woocommerce-admin/tree/main/docs/examples/extensions/add-task). You can do it by executing something like: `npm run example -- --ext=add-task` in your terminal
2. Go to the `Home` screen.
3. Hide the core task list.
Screen Capture on 2021-01-12 at 15-47-45
4. The task list will still be there.

**Expected behavior**
The list should disappear after pressing `hide`.

### PR DESCRIPTION
**Pull request by octaedro**
Fixes #6049

This PR fixes the error with hiding the core task list. After adding an extended task list, it's not possible to hide the core task list.

_Related also to [issue 6023](https://github.com/woocommerce/woocommerce-admin/issues/6023)_

### Screenshots
Screen Capture on 2021-01-12 at 15-56-42

### Detailed test instructions:

1. Add [a task list item](https://github.com/woocommerce/woocommerce-admin/tree/main/docs/examples/extensions/add-task). You can do it by executing something like: `npm run example -- --ext=add-task` in your terminal
2. Go to the `Home` screen.
3. Hide the core task list.
4. The task list should disappear.

- Also, go to `/wp-admin/admin.php?page=wc-admin&task=payments` and verify the _setup additional payment providers_ card is still visible.
screenshot-one wordpress test-2021 01 12-16_05_51

### CODE REVIEW

File: client/task-list/test/index.js [hunk 1]
```diff
@@ -135,6 +135,24 @@ describe( 'TaskDashboard and TaskList', () => {
 		).toBeNull();
 	} );
 
+	it( 'renders only the extended task list', () => {
```
Comment (louwie17) [inline client/task-list/test/index.js:138]: Could we also add a test for when the query object has `task` defined?

**Review by louwie17 — CHANGES_REQUESTED** [review]
Looks great @octaedro, and thanks for keeping check with the original change that had broken this - https://github.com/woocommerce/woocommerce-admin/pull/6026 
In light of that, it be great if we could have one more test case for when `task` is defined in the query object prop.

**Review by joshuatf — COMMENTED** [review]
Nice fix, @octaedro!  This PR is testing well; on a somewhat related note to this PR, I wanted to get your thoughts on https://github.com/woocommerce/woocommerce-admin/pull/5975#issuecomment-752989389

It looks like the original designs for the extended task list were in the original task list (which was previously extensible).  Did these designs change or did you run into trouble extending the original task list?

I ask because this introduces quite a few semantic changes around the task list and I can see more issues like this popping up going forward.

File: client/task-list/test/index.js (hunk 1 above)
Comment (octaedro) [reply client/task-list/test/index.js:158]: Good idea! I added the test in the commit b62f9cf

**Comment by octaedro** [comment]
Thank you @joshuatf for your review.

> This PR is testing well; on a somewhat related note to this PR, I wanted to get your thoughts on #5975 (comment)

My plan is to refactor a few things (swap the `do_action` and `add_action` and address some modifications suggested [here](https://github.com/woocommerce/woocommerce-admin/pull/5833#pullrequestreview-560540018)) in a follow-up of [this PR](https://github.com/woocommerce/woocommerce-admin/pull/5833).

> It looks like the original designs for the extended task list were in the original task list (which was previously extensible). Did these designs change or did you run into trouble extending the original task list?

Yes, the design was changed, now we're showing the core list and the extended list apart, for more context here is [the design](https://github.com/woocommerce/woocommerce-admin/issues/5242#issuecomment-733171105) and [the issue](https://github.com/woocommerce/woocommerce-admin/issues/5242).

> I ask because this introduces quite a few semantic changes around the task list and I can see more issues like this popping up going forward.

For sure if you have changes or suggestions please let me know and I'll be happy to address them 🙂

File: client/task-list/test/index.js (hunk 1 above)
Comment (louwie17) [reply client/task-list/test/index.js:158]: Thanks @octaedro, looks great :)

**Review by louwie17 — APPROVED** [review]
Thanks for adding the extra test case, LGTM 🚀

**Comment by joshuatf** [comment]
>Yes, the design was changed, now we're showing the core list and the extended list apart, for more context here is the design and the issue.

Gotcha.  Thanks for that context @octaedro.  It seems like the idea is to make sure extensions don't add items under the main task list, is that right?  I'm curious if this is an area that's being walled off intentionally or not (/cc @elizaan36).

If the idea is to intentionally wall off the list, extensions are still able to add tasks via the `woocommerce_admin_onboarding_task_list` filter and we also have an example plugin that shows how to do this.

>My plan is to refactor a few things (swap the do_action and add_action and address some modifications suggested here) in a follow-up of this PR.

Awesome!  A couple things that stood out to me were differences in the logic in a few places.  For example, in some places, this is the `taskListHidden` logic:

```js
const taskListHidden = getOption( 'woocommerce_task_list_hidden' ) === 'yes';
```

While in others it is:

```js
taskListHidden:
	getOption( 'woocommerce_task_list_hidden' ) === 'yes' &&
	getOption( 'woocommerce_extended_task_list_hidden' ) !== 'no',
```

It might be good to create a utils file that houses this logic so we have a single source of truth.

On the server we use the logic:

```php
	public static function should_show_tasks() {
		return 'no' === get_option( 'woocommerce_task_list_hidden', 'no' ) || 'no' === get_option( 'woocommerce_extended_task_list_hidden', 'no' );
	}
```

What's also a bit confusing about all this naming is the semantics of this.  For example, maybe these should be renamed to `taskListsHidden` (note the extra `s`) and rename any variables to differentiate the original task list from the extensions one.

Thanks again for all your work on this!

**Comment by elizaan36** [comment]
>  It seems like the idea is to make sure extensions don't add items under the main task list, is that right? I'm curious if this is an area that's being walled off intentionally or not (/cc @elizaan36).

The main problem to solve is that extensions will add tasks after core setup is complete. So we'd run into the problem of having a completed and dismissed setup task list, then an extension adds a task, and the task list would have to reappear despite the user choice to hide it.

So it's really about making sure extension tasks can operate independently of the setup task list, as they'll be associated with different points in the setup/store management timeline.

> If the idea is to intentionally wall off the list, extensions are still able to add tasks via the woocommerce_admin_onboarding_task_list filter and we also have an example plugin that shows how to do this.

I think it could be problematic if we still allow extensions to add tasks to the store setup task list for the reason above. One solution could be, and this may get super in the weeds, but we could classify the extensions installed in the onboarding wizard differently so they have the ability to show up in the setup task list.

**Comment by octaedro** [comment]
Thank you @joshuatf for your suggestions.
I created the [PR 6081](https://github.com/woocommerce/woocommerce-admin/pull/6081) addressing them. I also added a few modifications that Jeff [proposed here](https://github.com/woocommerce/woocommerce-admin/pull/5833#pullrequestreview-560540018).

**Pull request merged**